app/src/main/java/com/example/happybirthday/unit2/UnitTwoActivity.kt

```diff
@@ -1,11 +1,42 @@
 package com.example.happybirthday.unit2
 
+
 import android.os.Bundle
 import androidx.appcompat.app.AppCompatActivity
+import com.example.happybirthday.R
+import com.example.happybirthday.databinding.ActivityMainBinding
+import java.text.NumberFormat
 
 class UnitTwoActivity : AppCompatActivity() {
 
+    lateinit var binding: ActivityMainBinding
+
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+        binding = ActivityMainBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+
+        binding.calculateTips.setOnClickListener { calculateTip() }
     }
-}
\ No newline at end of file
+
+    private fun calculateTip() {
+        val textValue = binding.costOfService.text.toString()
+        val cost = textValue.toDoubleOrNull()
+        val tipsPercentage = when (binding.tipsOptions.checkedRadioButtonId) {
+            R.id.twenty_percent_option -> 0.20
+            R.id.eighteen_percent_option -> 0.18
+            else -> 0.15
+        }
```

Comment:
It's better to always write full `when`. It could be a potential bug here, if you add a new option for example `ten_percent_option`, but forget to change this `when`, it will be treated as `0.15`. But if we throw an exception, we will definitely find the bug.
```
val tipsPercentage = when (val id = binding.tipsOptions.checkedRadioButtonId) {
            R.id.twenty_percent_option -> 0.20
            R.id.eighteen_percent_option -> 0.18
            R.id.fifteen_percent_option -> 0.15
            else -> throw IllegalArgumentException("unsupported id: $id")
        }
```